### ISSUE TITLE
Why I cannot get the first mouse-down event on focused window.

### ISSUE THREAD

**Issue by ChivenZhang** [ocornut/imgui#6509]
**Version/Branch of Dear ImGui:**

Version: v1.89.4 WIP
Branch: docking

**Back-end/Renderer/Compiler/OS**

Back-ends: imgui_impl_opengl3.cpp + imgui_impl_SDL2.cpp
Compiler: msvc 2019
Operating System: windows 11

**My Issue/Question:**

I would like to fetch the mouse-down event when I focused the window by clicking mouse button. But there was something wierd happend when I moved the focus from window A to window B, I got the mouse-down event in window A. It's not correct. It's my test code below, can anybody tell me why? 
Thanks.

```
static bool show_test_window = true;
if (show_test_window)
{
    ImGui::SetNextWindowSize({ 200,200 }, ImGuiCond_Appearing);
    ImGui::Begin("Test A", &show_test_window);
    if (ImGui::IsWindowFocused(ImGuiFocusedFlags_ChildWindows))
    {
        for (auto i = 0; i < ImGuiMouseButton_COUNT; i++)
        {
            if (ImGui::IsMouseDown(i) && io.MouseDownDuration[i] == 0.0f)
            {
                printf("mouse down in A\n");
            }
        }
    }
    ImGui::End();

    ImGui::SetNextWindowSize({ 200,200 }, ImGuiCond_Appearing);
    ImGui::Begin("Test B", &show_test_window);
    if (ImGui::IsWindowFocused(ImGuiFocusedFlags_ChildWindows))
    {
        for (auto i = 0; i < ImGuiMouseButton_COUNT; i++)
        {
            if (ImGui::IsMouseDown(i) && io.MouseDownDuration[i] == 0.0f)
            {
                printf("mouse down in B\n");
            }
        }
    }
    ImGui::End();
}
```

**Comment by ocornut** [comment]
Focusing is happening at the end of the frame.

> I would like to fetch the mouse-down event when I focused the window by clicking mouse button.

This is likely a [XY Problem](https://xyproblem.info/) so I would suggest that you explain in more details what you are trying to achieve and why, and we can provide a better answer.

Please note that window focusing can happen from gamepad, keyboard or API calls at all. So unless you explicitly need to test for mouse only focus you may not be heading in the right direction.

PS: for reference: `IsMouseClicked(i)` is equivalent to `mGui::IsMouseDown(i) && io.MouseDownDuration[i] == 0.0f`)

**Comment by ChivenZhang** [comment]
> Focusing is happening at the end of the frame.
> 
> > I would like to fetch the mouse-down event when I focused the window by clicking mouse button.
> 
> This is likely a [XY Problem](https://xyproblem.info/) so I would suggest that you explain in more details what you are trying to achieve and why, and we can provide a better answer.
> 
> Please note that window focusing can happen from gamepad, keyboard or API calls at all. So unless you explicitly need to test for mouse only focus you may not be heading in the right direction.
> 
> PS: for reference: `IsMouseClicked(i)` is equivalent to `mGui::IsMouseDown(i) && io.MouseDownDuration[i] == 0.0f`)

I want to pass io event for every specific window, such as mouse/key/gameboard device. So it's important to ensure the right order when user input actions hanppen.

**Comment by ocornut** [comment]
I don't understand your answer it is very ambiguous. Please be more detailed and explicit.

**Comment by ChivenZhang** [comment]
![image](https://github.com/ocornut/imgui/assets/48938612/6bf1f35c-2343-48c3-b0f3-f89953e0b9c6)
This is what I have done to wrap code to object-orientied style. I need to ensure mouse events be passed to plugin module when user click the window. The problem is when user clicked the window A, the event happend in window B. It's not usual when clicking the window more times but clicking window once from another one.

**Comment by ChivenZhang** [comment]
If I remove the condition "IsWindowFocused", the mouse click event would happen in every window, not the specific one. Is there a better way to identify the mouse event source ?

**Comment by ChivenZhang** [comment]
![image](https://github.com/ocornut/imgui/assets/48938612/ce43e9c4-ff10-4a6d-9674-aea7481da8be)
As the picture shows, when I clicked the window B, the focus moved from A to B, but the log indicated the event happened in window A. It seems that I got the result from the last step.

**Comment by ocornut** [comment]
It might make better sense to fill the window with InvisibleButton() and use IsItemHovered() || IsItemActive() for your forward logic.

Clicking window empty space on frame N will always focus window at the end of the frame.

**Comment by ChivenZhang** [comment]
I worked out a good solution about window focus event happening behind mouse button. I canceled the mouse event passing when window not holding the focus in current frame. The code looks like below:
```
static bool show_test_window = true;
static bool have_got_focus = false;
if (show_test_window)
{
    ImGui::SetNextWindowSize({ 200,200 }, ImGuiCond_Appearing);
    ImGui::Begin("Test A", &show_test_window);
    if(have_got_focus)
    {
        for (auto i = 0; i < ImGuiMouseButton_COUNT; i++)
        {
            if (ImGui::IsMouseClicked(i)) printf("mouse down in A\n");
        }
    }
    if (ImGui::IsWindowFocused(ImGuiFocusedFlags_ChildWindows)) have_got_focus = true;
    else have_got_focus = false;
    ImGui::End();
}
```

**Comment by ChivenZhang** [comment]
I have discarded the first mouse-down event before the window obtain focus at the end of frame. So It may be solved. Thanks
